dask/array/creation.py

```diff
@@ -245,6 +249,122 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, chunks='auto',
+    *args :
+        Other positional arguments to pass to function. Values must be constants
+        (not dask.arrays).
+    shape : Tuple[int], optional
+        Shape of the resulting array.
+    chunks : tuple, optional
+        Chunk shape of resulting blocks. If not provided, a chunking scheme
+        is chosen automatically.
+    dtype : np.dtype, optional
+        The ``dtype`` of the output array. It is recommended to provide this.
+        If not provided, will be inferred by applying the function to a small
+        set of fake data. For that call, the `dtype` member of `block_info`
+        will be ``None``.
+    name : str, optional
+        The key name to use for the output array. If not provided,
+        will be determined from `func`.
+    **kwargs :
+        Other keyword arguments to pass to function. Values must be constants
+        (not dask.arrays)
+
+    Examples
+    --------
+    This is a simplified version of :func:`eye` which only handles square
+    arrays with the ones on the main diagonal.
+
+    >>> def eye_chunk(block_info):
+    ...     r0, r1 = block_info.array_location[0]
+    ...     c0, c1 = block_info.array_location[1]
+    ...     if r0 == c0:
+    ...         return np.eye(r1 - r0, c1 - c0, dtype=block_info.dtype)
+    ...     else:
+    ...         return np.zeros(block_info.chunk_shape, dtype=block_info.dtype)
+    >>> from_block_function(eye_chunk, shape=(4, 4), chunks=2, dtype=float)
+    dask.array<eye_chunk, shape=(4, 4), dtype=float64, chunksize=(2, 2)>
+    >>> _.compute()
+    array([[1., 0., 0., 0.],
+           [0., 1., 0., 0.],
+           [0., 0., 1., 0.],
+           [0., 0., 0., 1.]])
+    """
+
+    shape = kwargs.pop('shape', None)
+    chunks = kwargs.pop('chunks', 'auto')
+    dtype = kwargs.pop('dtype', None)
+    name = kwargs.pop('name', None)
+    name = '%s-%s' % (name or funcname(func), tokenize(func, shape, dtype, chunks, args, kwargs))
+
+    if dtype is None:
+        chunks = normalize_chunks(chunks, shape)
+        dummy_block_info = FromBlockFunctionInfo(
+            shape, shape, (0,) * len(shape), [(0, 1)] * len(shape), (1,) * len(shape), None)
+        dtype = apply_infer_dtype(func, (dummy_block_info,) + args, kwargs, 'from_block_function')
+    else:
+        dtype = np.dtype(dtype)
+        chunks = normalize_chunks(chunks, shape, dtype=dtype)
+
+    # Allow user to omit shape when chunks are given in normalized form
+    shape = tuple(sum(bd) for bd in chunks)
+
+    keys = list(product([name], *[range(len(bd)) for bd in chunks]))
+    aggdims = [list(accumulate(add, (0,) + bd)) for bd in chunks]
+    locdims = [list(zip(a[:-1], a[1:])) for a in aggdims]
+    locations = list(product(*locdims))
+    num_chunks = tuple(len(bd) for bd in chunks)
+    args = quote(args)
+    dsk = {key: (_call_from_block_function, func, shape, num_chunks, key[1:], location, dtype, args, kwargs)
+           for key, location in zip(keys, locations)}
```

Comment:
Also, this is probably out of scope, but it would be very valuable if this graph layer was a `Blockwise` object.  This would allow task fusion down to our creation functions, which would be highly valuable.  This probably isn't trivial though, and may require some modifications to Blockwise itself.